tuplex/core/src/logical/UDFOperator.cc

```diff
@@ -63,57 +64,68 @@ namespace tuplex {
-                        return Schema::UNKNOWN;
-                    } else {
-                        // check what return type was given
-                        auto output_type = _udf.getOutputSchema().getRowType();
-                        if(output_type.isExceptionType()) {
+                    if(_udf.getTypeError().empty()) {
+                        if(!success) {
+                            Logger::instance().defaultLogger().info("was not able to detect type for UDF, statically and via tracing."
+                                                                    " Provided parent row type was " + parentSchema.getRowType().desc() );
+                            // 4. mark as uncompilable UDF, but sample output row type, so at least subsequent operators
+                            //    can get compiled!
                             _udf.markAsNonCompilable();
-                            // exception type? => sample produced only exceptions! warn user.
-                            Logger::instance().defaultLogger().error("Tracing via sample produced mostly"
-                                                                     " exceptions, please increase sample size"
-                                                                     " or alter UDF to not yield exceptions only.");
-                            // @TODO: print nice traceback with sample...
-                            // => should use beefed up sample processor class for this...
+                            throw std::runtime_error("TODO: implement via sampling of Python UDF executed most likely type, so pipeline can still get compiled.");
                             return Schema::UNKNOWN;
                         } else {
-                            // all good, keep sampled type but mark as non compilable
+                            // check what return type was given
+                            auto output_type = _udf.getOutputSchema().getRowType();
+                            if(output_type.isExceptionType()) {
+                                _udf.markAsNonCompilable();
+                                // exception type? => sample produced only exceptions! warn user.
+                                Logger::instance().defaultLogger().error("Tracing via sample produced mostly"
+                                                                         " exceptions, please increase sample size"
+                                                                         " or alter UDF to not yield exceptions only.");
+                                // @TODO: print nice traceback with sample...
+                                // => should use beefed up sample processor class for this...
+                                return Schema::UNKNOWN;
+                            } else {
+                                // all good, keep sampled type but mark as non compilable
```

Comment:
what does this comment mean? where is the udf being marked non-compilable in this case?

Reply:
means the function can't get compiled to LLVM IR, but an output type was detected so subsequent functions can get compiled. I.e., this particular function will be executed inline as compiled code.

Reply:
@rahulyesantharao: Any suggestion how we could reformulate this better?

Reply:
oh I see, I think reading the full 3-stage typing made it clearer. Maybe just a comment saying what you just said here? i.e. "Cannot statically type AST, but sampling yields common-case output type to propagate to subsequent stages".

Reply:
any updates on this?

Reply:
I added a comment what @rahulyesantharao suggested.